code/tracing/tracing.cpp

```diff
@@ -103,507 +70,299 @@ static int get_query_object() {
- */
-void profile_end(const char* name)
-{
-	if (Cmdline_json_profiling)
-	{
-		std::lock_guard<std::mutex> guard(json_mutex);
+void shutdown() {
+	while (!gpu_events.empty()) {
+		process_events();
 
-		tracing_data data;
+		// Don't do busy waiting...
+		os_sleep(5);
+	}
 
-		data.name = name;
+	for (auto query : query_objects) {
+		gr_delete_query_object(query);
+	}
+	query_objects.clear();
 
-		data.pid = get_pid();
-		data.tid = get_tid();
+	mainFrameTimer = nullptr;
+	traceEventWriter = nullptr;
 
-		data.enter = false;
-		data.time = timer_get_nanoseconds();
+	initialized = false;
+}
 
-		if (do_gpu_queries) {
-			data.gpu_query = get_query_object();
-			gr_query_value(data.gpu_query, QueryType::Timestamp);
-		} else {
-			data.gpu_query = -1;
-		}
+namespace complete {
 
-		current_frame_data.push_back(data);
+void start(const Category& category, trace_event* evt) {
+	if (!do_trace_events) {
+		// No one to process the event is here
+		return;
 	}
 
-	if (Cmdline_frame_profile) {
-		int num_parents = 0;
-		int child_of = -1;
-
-		for ( int i = 0; i < (int)samples.size(); i++ ) {
-			if ( samples[i].open_profiles ) {
-				if ( samples[i].num_children == 1 ) {
-					child_of = i;
-				}
-			}
-		}
+	if (!initialized) {
+		return;
+	}
 
-		for ( int i = 0; i < (int)samples.size(); i++ ) {
-			if ( !strcmp(samples[i].name.c_str(), name) && samples[i].parent == child_of ) {
-				int inner = 0;
-				int parent = -1;
-				std::uint64_t end_time = timer_get_microseconds();
-				samples[i].open_profiles--;
-
-				// count all parents and find the immediate parent
-				while ( inner < (int)samples.size() ) {
-					if ( samples[inner].open_profiles > 0 ) {
-						// found a parent (any open profiles are parents)
-						num_parents++;
-
-						if (parent < 0) {
-							// replace invalid parent (index)
-							parent = inner;
-						}
-						else if (samples[inner].start_time >= samples[parent].start_time) {
-							// replace with more immediate parent
-							parent = inner;
-						}
-					}
-					inner++;
-				}
+	init_event(category, evt);
 
-				// remember the current number of parents of the sample
-				samples[i].num_parents = num_parents;
+	evt->duration = 0;
+	evt->type = EventType::Complete;
 
-				if ( parent >= 0 ) {
-					// record this time in children_sample_time (add it in)
-					samples[parent].children_sample_time += end_time - samples[i].start_time;
-				}
+	if (do_gpu_queries && category.usesGPUCounter()) {
+		Assertion(get_tid() == main_thread_id, "This function must be called from the main thread!");
 
-				// save sample time in accumulator
-				samples[i].accumulator += end_time - samples[i].start_time;
+		gpu_trace_event gpu_event;
+		gpu_event.base_evt.category = &category;
+		gpu_event.base_evt.tid = 1;
+		gpu_event.base_evt.pid = GPU_PID;
+		gpu_event.base_evt.type = EventType::Begin;
 
-				break;
-			}
-		}
+		gpu_event.gpu_begin_query = get_gpu_timestamp_query();
```

Comment:
The GPU queries are inserted into the command stream at the moment when the `glQuery` is executed but there is no queue in the graphics system to be flushed so I'm not sure what you mean by that.

The problem with different time bases is that the TraceEvent viewer cannot handle them since it thinks all times originate from the same source so I need to synchronize them to avoid crashing the viewer. I hope that the timer in the GPU is ticking at the same rate as the CPU timer but if it isn't then this is the best result we can get without going into clock synchronization and clock drift compensation territory.